### PR TITLE
Refactor DPOS-User to extend Gateway-User

### PR DESCRIPTION
**Pull request by gakonst**
Extends https://github.com/loomnetwork/loom-js/pull/241

Refactor of code (backwards compatible) to allow dpos-user to be an extension of gateway-user which itself is an extension of crosschain. Crosschain has all account mapping related bindings, gateway-user all gatewya deposit/withdrawals, and dpos-user in addition has all the dpos features we use.

### CODE REVIEW

File: src/dpos-user.ts [hunk 1]
```diff
@@ -19,38 +20,13 @@ import {
...
-import { ValidatorManagerContract } from './mainnet-contracts/ValidatorManagerContract'
-import { ERC20Gateway_v2 } from './mainnet-contracts/ERC20Gateway_v2'
 import { LocalAddress } from './address'
 import { getMetamaskSigner } from './solidity-helpers'
 
-enum GatewayVersion {
-  SINGLESIG = 1,
-  MULTISIG = 2
-}
+import { GatewayUser } from './gateway-user'
```
Comment (enlight) [inline src/dpos-user.ts]: These imports should all be up top, and there's already an import from `./gateway-user` up there

File: src/gateway-user.ts [hunk 2]
```diff
@@ -0,0 +1,358 @@
...
+
+import { CrossChain } from './crosschain'
+
+const log = debug('dpos-user')
+
+const coinMultiplier = new BN(10).pow(new BN(18))
+
+import { ERC20 } from './mainnet-contracts/ERC20'
+import { ValidatorManagerContract } from './mainnet-contracts/ValidatorManagerContract'
+import { ERC20Gateway_v2 } from './mainnet-contracts/ERC20Gateway_v2'
```
Comment (enlight) [inline src/gateway-user.ts]: Imports up top

File: src/crosschain.ts [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+import Web3 from 'web3'
+
+import { createDefaultClient } from './helpers'
+import { Address, LocalAddress, Client, Contracts, EthersSigner } from '.'
+
+import { AddressMapper } from './contracts/address-mapper'
+
+const log = debug('crosschain')
+
+export class CrossChain {
```
Comment (enlight) [inline src/crosschain.ts]: -> `CrossChainUser`
And `crosschain.ts` -> `crosschain-user.ts`

File: src/crosschain.ts [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+    dappchainEndpoint: string,
+    dappchainKey: string,
+    chainId: string
+  ): Promise<CrossChain> {
+    const provider = new ethers.providers.JsonRpcProvider(endpoint)
+    const wallet = new ethers.Wallet(privateKey, provider)
+    return CrossChain.createCrossChainUserAsync(wallet, dappchainEndpoint, dappchainKey, chainId)
+  }
+
+  static async createMetamaskCrossChainUserAsync(
```
Comment (enlight) [inline src/crosschain.ts]: I think you need to merge in some of the eth-signer stuff from master.

File: src/gateway-user.ts [hunk 5]
```diff
@@ -0,0 +1,358 @@
...
+  get dappchainGateway(): Contracts.LoomCoinTransferGateway {
+    return this._dappchainGateway
+  }
+
+  /**
+   * Deposits funds from mainnet to the gateway
+   */
+  async depositAsync(amount: BN): Promise<ethers.ContractTransaction> {
+    let currentApproval = await this._ethereumLoom.functions.allowance(
+      await this.ethAddress,
```
Comment (enlight) [inline src/gateway-user.ts]: Why is there an `await` here?

File: src/gateway-user.ts [hunk 6]
```diff
@@ -0,0 +1,358 @@
...
+   * Deposits an amount of LOOM tokens to the dappchain gateway and return a signature which can be used to withdraw the same amount from the mainnet gateway.
+   *
+   * @param amount The amount that will be deposited to the DAppChain Gateway (and will be possible to withdraw from the mainnet)
+   */
+  private async depositCoinToDAppChainGatewayAsync(amount: BN): Promise<string> {
+    let pendingReceipt = await this.getPendingWithdrawalReceiptAsync()
+    let signature: Uint8Array
+    if (pendingReceipt === null) {
+      await this._dappchainLoom.approveAsync(this._dappchainGateway.address, amount)
+      const ethereumAddressStr = await this.ethAddress
```
Comment (enlight) [inline src/gateway-user.ts]: Why is there an `await` here?

File: src/gateway-user.ts (hunk 5 above)
Comment (gakonst) [reply src/gateway-user.ts]: All ethereum functions are async

File: src/gateway-user.ts (hunk 6 above)
Comment (gakonst) [reply src/gateway-user.ts]: Waiting for the approve promise to resolve, then move on

File: src/crosschain.ts (hunk 4 above)
Comment (gakonst) [reply src/crosschain.ts]: yeah will do on the base branch

Comment (gakonst) [reply src/crosschain.ts]: yeah will do on the base branch

**Pull request merged**